source/opt/decoration_manager.cpp

```diff
@@ -18,6 +18,93 @@
+using spvtools::ir::Operand;
+
+// Returns true if |inst1| is strictly less than |inst2|, for SpvOpDecorate and
+// SpvOpDecorateId.
+//
+// |inst1| is considered to be strictly less than |inst2| if:
+// * the decoration of |inst1| is strictly less than the one from |inst2|;
+// * and |inst1| has strictly fewer operands than |inst2|;
+// * and when comparing operands
+//   * |inst1|'s operand has strictly fewer words than |inst2|'s one;
+//   * and each word of |inst1|'s operand is strictly less than the ones from
+//     |inst2| (using their representation as uint32_t).
+bool compareBeforeDecorate(const Instruction* inst1, const Instruction* inst2) {
+  if (inst1->GetSingleWordInOperand(1u) > inst2->GetSingleWordInOperand(1u))
+    return false;
+  if (inst1->NumInOperands() > inst2->NumInOperands()) return false;
+  for (uint32_t i = 2u; i < inst1->NumInOperands(); ++i) {
+    const Operand& op1 = inst1->GetInOperand(i);
+    const Operand& op2 = inst2->GetInOperand(i);
+    if (op1.words.size() > op2.words.size()) return false;
+    for (uint32_t j = 0u; j < op1.words.size(); ++j)
+      if (op1.words[j] > op2.words[j]) return false;
+  }
+  return true;
+}
+
+// Returns true if |inst1| is strictly less than |inst2|, for
+// SpvOpMemberDecorate.
+//
+// |inst1| is considered to be strictly less than |inst2| if:
+// * the decoration of |inst1| is strictly less than the one from |inst2|;
+// * the member index of |inst1| is strictly less than the one from |inst2|;
+// * and |inst1| has strictly fewer operands than |inst2|;
+// * and when comparing operands
+//   * |inst1|'s operand has strictly fewer words than |inst2|'s one;
+//   * and each word of |inst1|'s operand is strictly less than the ones from
+//     |inst2| (using their representation as uint32_t).
+bool compareBeforeMemberDecorate(const Instruction* inst1,
+                                 const Instruction* inst2) {
+  if (inst1->GetSingleWordInOperand(2u) > inst2->GetSingleWordInOperand(2u))
+    return false;
+  if (inst1->GetSingleWordInOperand(1u) > inst2->GetSingleWordInOperand(1u))
+    return false;
+  if (inst1->NumInOperands() > inst2->NumInOperands()) return false;
+  for (uint32_t i = 3u; i < inst1->NumInOperands(); ++i) {
+    const Operand& op1 = inst1->GetInOperand(i);
+    const Operand& op2 = inst2->GetInOperand(i);
+    if (op1.words.size() > op2.words.size()) return false;
+    for (uint32_t j = 0u; j < op1.words.size(); ++j)
+      if (op1.words[j] > op2.words[j]) return false;
+  }
+  return true;
+}
+
+// Returns true if |inst1| is equal to |inst2|, for SpvOpDecorate and
+// SpvOpDecorateId.
+//
+// When checking for equality, the target is not considered. The actual check
+// is the same as done by compareBeforeDecorate, but replacing all instances of
+// "strictly less than" by "equal to".
+bool compareEqualDecorate(const Instruction* inst1, const Instruction* inst2) {
+  if (inst1->NumInOperands() != inst2->NumInOperands()) return false;
```

Comment:
Consider comparing all words after target instead of operands.

Reply:
The `bool operator!=(const Operand& op1, const Operand& op2)` relies on the equal operator of Operand, which does a word to word comparison already (as it uses the == operator of std::vector).